src/engine/IndexScan.cpp

```diff
@@ -324,15 +341,192 @@ Permutation::IdTableGenerator IndexScan::lazyScanForJoinOfColumnWithScan(
+                   "num-blocks-skipped-graph");
+  updateIfPositive(metadata.numBlocksPostprocessed_,
+                   "num-blocks-postprocessed");
+  updateIfPositive(metadata.numBlocksWithUpdate_, "num-blocks-with-update");
+}
+
+// Store a Generator and its corresponding iterator as well as unconsumed values
+// resulting from the generator.
+struct IndexScan::SharedGeneratorState {
+  Result::Generator generator_;
+  ColumnIndex joinColumn_;
+  Permutation::MetadataAndBlocks metaBlocks_;
+  std::optional<Result::Generator::iterator> iterator_ = std::nullopt;
+  std::optional<Id> lastValue_ = std::nullopt;
+  std::deque<Result::IdTableVocabPair> prefetchedValues_{};
+  std::vector<CompressedBlockMetadata> pendingBlocks_{};
+  bool hasUndef_ = false;
+  bool doneFetching_ = false;
+
+  // Fetch the first value from the generator and skip empty tables.
+  bool fetchFirst() {
+    AD_CORRECTNESS_CHECK(!iterator_.has_value());
+    iterator_ = generator_.begin();
+    while (iterator_ != generator_.end()) {
+      if (iterator_.value()->idTable_.empty()) {
+        ++iterator_.value();
+        continue;
+      }
+      hasUndef_ = iterator_.value()->idTable_.at(0, joinColumn_).isUndefined();
+      return false;
+    }
+    return true;
+  }
+
+  // Consume the next non-empty table from the generator and calculate the next
+  // matching blocks from the index scan.
+  void fetch() {
+    bool exhausted;
+    if (iterator_.has_value()) {
+      AD_CONTRACT_CHECK(iterator_.value() != generator_.end());
+      do {
+        // Skip empty tables
+        ++iterator_.value();
+        exhausted = iterator_.value() == generator_.end();
+      } while (!exhausted && iterator_.value()->idTable_.empty());
+    } else {
+      exhausted = fetchFirst();
+    }
+    if (!exhausted) {
+      prefetchedValues_.push_back(std::move(*iterator_.value()));
+      auto joinColumn =
+          prefetchedValues_.back().idTable_.getColumn(joinColumn_);
+      AD_EXPENSIVE_CHECK(std::ranges::is_sorted(joinColumn));
+      // Skip processing for undef case, it will be handled differently
+      if (joinColumn.empty() || !joinColumn[0].isUndefined()) {
+        AD_CORRECTNESS_CHECK(!hasUndef_);
+        // Find first value that differs from the last one that was used to find
+        // matching blocks.
+        size_t offset = std::ranges::find_if_not(joinColumn,
+                                                 [this](const auto& element) {
+                                                   return element == lastValue_;
+                                                 }) -
+                        joinColumn.begin();
+        auto newValues = joinColumn.subspan(offset);
+        if (!newValues.empty()) {
+          std::ranges::move(CompressedRelationReader::getBlocksForJoin(
+                                newValues, metaBlocks_),
+                            std::back_inserter(pendingBlocks_));
+          lastValue_ = newValues.back();
```

Comment:
I don't this that this is correct.
If the first input block contains only the value `1` and the second input block the value `2` , and there is a block that contains ids in the range e.g. [0, 5], then that block will be yielded twice, which is not correct.
You have to at least exclude all previously yielded scan blocks from the further `getBlocksForJoin` cases.

A further optimization would be to also exclude blocks, that are already provably too small (s.t. you make guaranteed progress in the scan blocks for each input block.

Reply:
And if an input block doesn't match any of the scan blocks, you can also immediately drop it here
(small optimization, but hopefully easy to implement).